### ISSUE TITLE
C++ Pub/Sub stops receiving

### ISSUE THREAD

**Issue by BrianPez** [googleapis/google-cloud-cpp#9171]
Using the online documentation and the code in the sample repo, I've implemented PubSub for a single incoming and multiple outgoing.  The pull PubSub is an ordering subscription (would much prefer an ordered and deliever once, but that doesn't seem allowed).  I have dead letter configured for each PubSub.  Repeatably, the code will stop pulling messages.  Sometimes in 5 minutes, sometimes after a couple of days.  This morning, there were 55k messages waiting to be pulled.  No messages in the dead letter.  Based on the website, I expected something from the then() handler in the subscription, but there is no error messages.  Here is my code.

```c++
/* These are based on https://github.com/googleapis/google-cloud-cpp/blob/HEAD/google/cloud/pubsub/samples/samples.cc */
/* https://cloud.google.com/pubsub/docs/handling-failures#dead-letter_topic */
/* https://cloud.google.com/blog/products/application-development/running-cplusplus-apps-with-the-help-of-pubsub-and-gke */
/* https://cloud.google.com/pubsub/docs/samples/pubsub-subscriber-error-listener#code-sample */
#include <google/cloud/pubsub/publisher.h>
#include <google/cloud/storage/client.h>
#include <google/cloud/pubsub/subscriber.h>
#include <google/cloud/pubsub/subscription_admin_client.h>
#include <google/cloud/pubsub/topic_admin_client.h>
#include <nlohmann/json.hpp>

    void createTopic(const std::string& topicId, const std::string& dlTopicId, google::cloud::pubsub::TopicAdminClient& client)
    {
        namespace pubsub = ::google::cloud::pubsub;

        [](pubsub::TopicAdminClient client, std::string project_id, std::string topic_id, std::string dl_topic_id)
            {
                auto topic = client.CreateTopic(pubsub::TopicBuilder(pubsub::Topic(project_id, topic_id)));
                // Note that kAlreadyExists is a possible error when the library retries.
                if (topic.status().code() == google::cloud::StatusCode::kAlreadyExists)
                {
                    ;
                }
                else if (!topic)
                {
                    throw std::runtime_error(topic.status().message());
                }
                else
                {
                    LOGn << "The topic " << topic_id << " was successfully created: " << topic->DebugString();
                }

                auto dl_topic = client.CreateTopic(pubsub::TopicBuilder(pubsub::Topic(project_id, dl_topic_id)));
                // Note that kAlreadyExists is a possible error when the library retries.
                if (dl_topic.status().code() == google::cloud::StatusCode::kAlreadyExists)
                {
                    ;
                }
                else if (!dl_topic)
                {
                    throw std::runtime_error(dl_topic.status().message());
                }
                else
                {
                    LOGn << "The deadletter topic " << dl_topic_id << " was successfully created: " << dl_topic->DebugString();
                }

            }
        (client, pubSubProjectId, topicId, dlTopicId);
    }
    void createOrderingSubscription(const std::string& topicId, const std::string& subscriptionId, const std::string& dlTopicId, const std::string& dlSubscriptionId, google::cloud::pubsub::SubscriptionAdminClient& client)
    {
        namespace pubsub = ::google::cloud::pubsub;

        return [](pubsub::SubscriptionAdminClient client, std::string const& project_id, std::string const& topic_id, std::string const& subscription_id, std::string const& dl_topic_id, std::string const& dl_subscription_id)
            {
                static const constexpr int DEAD_LETTER_DELIVERY_ATTEMPTS = 5;

                auto sub = client.CreateSubscription(
                    pubsub::Topic(project_id, std::move(topic_id)),
                    pubsub::Subscription(project_id, subscription_id),
                    pubsub::SubscriptionBuilder{}
                        .enable_message_ordering(true)
                        .set_ack_deadline(std::chrono::seconds(60))
                        .set_dead_letter_policy(
                            pubsub::SubscriptionBuilder::MakeDeadLetterPolicy(
                            pubsub::Topic(project_id, dl_topic_id),
                            DEAD_LETTER_DELIVERY_ATTEMPTS))
                );
                if (sub.status().code() == google::cloud::StatusCode::kAlreadyExists)
                {
                    ;
                }
                else if (!sub)
                {
                    throw std::runtime_error(sub.status().message());
                }
                else
                {
                    LOGd << "The subscription " << subscription_id << " was successfully created: " << sub->DebugString();
                }

                auto dl_sub = client.CreateSubscription(
                    pubsub::Topic(project_id, std::move(dl_topic_id)),
                    pubsub::Subscription(project_id, dl_subscription_id)
                );
                if (dl_sub.status().code() == google::cloud::StatusCode::kAlreadyExists)
                {
                    ;
                }
                else if (! dl_sub)
                {
                    throw std::runtime_error(sub.status().message());
                }
                else
                {
                    LOGd << "The deadletter subscription " << dl_subscription_id << " was successfully created: " << dl_sub->DebugString();
                }

                return;
            }
        (client, pubSubProjectId, topicId, subscriptionId, dlTopicId, dlSubscriptionId);
    }
void run()
{
    while (running)
    {
        try
        {
            namespace pubsub = ::google::cloud::pubsub;;
            int concurrency = 1;

            pubsub::TopicAdminClient topicClient(pubsub::MakeTopicAdminConnection());
            pubsub::SubscriptionAdminClient subscriptionClient(pubsub::MakeSubscriptionAdminConnection());

            createTopicSubscriptionId(eId, fId, "", fTopicId, fSubscriptionId, fDLTopicId, fDLSubscriptionId);
            createTopic(fTopicId, fDLTopicId, topicClient);
            createOrderingSubscription(fTopicId, fSubscriptionId, fDLTopicId, fDLSubscriptionId, subscriptionClient);
            fSubscriptionProblem = false;
            auto const subscription = pubsub::Subscription(pubSubProjectId, fSubscriptionId);
            auto subscriber = pubsub::Subscriber(pubsub::MakeSubscriberConnection(
                subscription,
                pubsub::SubscriberOptions{}
                    .set_max_outstanding_messages(concurrency)
                    .set_max_outstanding_bytes(concurrency * 1024)
                    .set_max_concurrency(concurrency),
                pubsub::ConnectionOptions{}.set_background_thread_pool_size(concurrency))
            );
            auto session = subscriber.Subscribe([](pubsub::Message const& m, pubsub::AckHandler h) {
                ProcessMsg::CommTopic topic = static_cast<ProcessMsg::CommTopic>(std::stoi(m.attributes()["topic"]));
                bool g = true;
                std::string msg = m.data();

                LOGi << "Received message, id " << m.message_id() << ", attempt " << h.delivery_attempt() <<", ordering key '" << m.ordering_key() << "', topic " << ProcessMsg::commTopic(topic) << " (" << topic << ")";
                messageCallback(msg);
                std::move(h).ack();
            })
            .then([](::google::cloud::future<google::cloud::Status> f) {
                LOGe << "Subscription problem : " << f.get();
                fSubscriptionProblem = true;
            });
            if (! session.valid())
            {
                LOGe << "Subscribe session invalid";
                break;
            }

            while (running && ! fSubscriptionProblem)
            {
                if (outgoingQueue.empty())
                {
                    std::this_thread::sleep_for(std::chrono::milliseconds(500));
                    continue;
                }

                std::string gTopicId = "";
                std::string gSubscriptionId = "";
                std::string gDLTopicId = "";
                std::string gDLSubscriptionId = "";
                const OutgoingMsg omsg = outgoingQueue.front();

                createTopicSubscriptionId(0, 0, omsg.identifier, gTopicId, gSubscriptionId, gDLTopicId, gDLSubscriptionId);
                createTopic(gTopicId, gDLTopicId, topicClient);
                createOrderingSubscription(gTopicId, gSubscriptionId, gDLTopicId, gDLSubscriptionId, subscriptionClient);
                try
                {
                    auto publisher = pubsub::Publisher(pubsub::MakePublisherConnection(pubsub::Topic(pubSubProjectId, gTopicId)));
                    auto id = publisher.Publish(pubsub::MessageBuilder{}
                                                    .SetData(omsg.msg)
                                                    .SetAttribute("timestamp", std::to_string(time(NULL)))
                                                    .Build()).get();
                    if (! id)
                    {
                        LOGe << "Failed to publish message " << id.status().message();
                    }
                    else
                    {
                        LOGd << "Published message, id : " << *id << " (" << omsg.msg << ") to " << pubSubProjectId << "/" << gTopicId;
                        outgoingQueue.pop_front();
                    }
                }
                catch (std::exception const& e)
                {
                    LOGe << "Failed to publish message " << e.what();
                }
            }
            if (running)
            {
               LOGn << "GCP problem, cancelling session";
               session.cancel();
               LOGd << "Waiting 15 seconds";
               std::this_thread::sleep_for(std::chrono::seconds(15));
            }
        }
        catch (std::runtime_error &e)
        {
            LOGe << "RuntimeError exception : " << e.what();
        }
        catch(std::exception& e)
        {
            LOGe << "Exception : " << e.what();
        }
    }
}
```

**Comment by coryan** [comment]
Thanks for the detailed bug report.  We are trying to repro the problem locally.  Your code seems to be missing some context, but we can make reasonable (or what we think are reasonable) guesses about it.

As to "exactly once" semantics.  That is in the roadmap.

**Comment by coryan** [comment]
Also looks like a dup of https://stackoverflow.com/questions/72519418

**Comment by BrianPez** [comment]
if it helps, here is the metrics from the GCP console web page (is this available in the API?  I couldn't find if).  The drop occurred this morning when I restarted the process.
![metrics](https://user-images.githubusercontent.com/47791401/172205028-1f510ee8-feaf-452f-aee0-a9a6e9a954c5.png)

**Comment by coryan** [comment]
Thanks for the graphs.  You are correct, there is nothing in the Pub/Sub API to query this.  I expect (but I have never done this myself, so don't take this as gospel) that you could query this data via the monitoring APIs:

https://github.com/googleapis/google-cloud-cpp/tree/main/google/cloud/monitoring

Be aware that this is intended for monitoring (like producing the chart), not to control application behavior.

**Comment by coryan** [comment]
Also: I do not doubt you are seeing this problem, I just don't know why yet.

**Comment by coryan** [comment]
While you gave us a lot of code, there are some key pieces missing (probably in an attempt to keep the code size manageable).  To be able to repro this, I need to confirm a few things:

1. Why are you dynamically creating the topics and subscriptions?  Those are (more often than not) created administratively and are intended to be long lived.  It is not that it is wrong to create them dynamically, it is just that we could simplify the repro code a lot if they were pre-created.  Would it be Okay if I did that on the first pass?
2. Can you say something about the message rate received by this application? Just trying to make sure any repro is realistic.
3. Can you say something about the number of topics you are sending "to" from this code?  It should work for whatever size (withing reason), just want to make sure any repro is (again) realistic.
4. A `pubsub::Subscriber` already reconnects if the streaming RPC is disconnected, and you can control how many attempts are made to create the streaming using `pubsub::RetryPolicyOption`.  You could avoid that whole `while(running) { try {` loop by taking advantage of those.  Would it be Okay if I made that simplification in a repro?
5. It is very inefficient to create a `pubsub::Publisher` to send a single message.  You should consider keeping a hash map of `pubsub::Publisher`s around to reuse them.
6. How is the `outgoingQueue` populated?  Is that done in the `messageCallback()` call?
7. If so, why the extra queue?  I would have written something like:

```cc
    auto session = subscriber.Subscribe([](pubsub::Message m, pubsub::AckHandler h) {
        auto topic_id = GetTopicId(m);
        auto publisher = LookupPublisher(topic_id); // create on-demand if needed
        publisher.Publish(std::move(m)).then([h = std::move(h)](auto f) {
          auto id = f.get();
          if (!id) {
            LOGd << "Publish failed with " << id.status() << " nacking the message";
            std::move(h).nack();
            return;
          }
         LOGd << "Message successfully redirected to topic " << topic_id << " with id " << *id;
         std::move(h).ack();
    });
 ```

**Comment by BrianPez** [comment]
I confirmed this morning that when the inbound subscription is no longer pulling messages, the outbound publisher is still working.  I wasn't sure if the the entire API had issues or just the subscribe side.

In answer of your questions.
1) My system is designed to add customers without any user intervention.  Any time a new customer signs up, one of these processes starts automatically.  Same thing with the publishing.  The customer can add hardware at any time requiring a new publish topic.  I could refactor this to move this code elsewhere (probably in PHP), but the basic premise will still be the same.

2) My rates are varied based on the customer and I am seeing problems in both low and high volumes.  One of my busier sites is getting about 10k messages an hour.  One of my slower sites is about 200 an hour.  Both my high volume and low volume sites are having this issue.

3) Currently, my Publish topics are 1 or 2 per process.  But I have no upper limits on the Publish topics.  Hundreds is possible.

4) The outer while loop had 2 purposes.  First, to gracefully handle a process shutdown (my preference would have been to do a pull request when I wanted, rather than executing a lambda callback).  Second, detect a failure in the subscribe and re-open it (for example, if someone deletes the subscription).  I have no problem with changing the logic.  I'm finding the documentation to be not the best (I didn't see anything about RetryPolicyOption on the web page that discusses error handling) and the only sample code I could find is a single file that executes all the functions.

5) I had plans to implement something similar, but when the Subscribe issues started surfacing, I stopped working on optimizations.

6) This code runs in one thread.  I have a separate thread that populates the outgoing queue which this thread is watching to Publish.

7) This is a good optimization that I should implement.  I was migrating some existing messaging queue into PubSub.  Those used the same queue for both incoming and outgoing which made the extra queue necessary.

**Comment by hjtran** [comment]
Hi all, sorry to jump in but I'm seeing what I think looks like a similar issue though using the python client.
```
google-cloud-core==2.2.3
google-cloud-pubsub==2.11.0
```
My subscriber workers also work for some amount of time before eventually stopping despite there being many messages left.

```
def _get_subscriber_client():
    global SUBSCRIBER
    if SUBSCRIBER is None:
        SUBSCRIBER = pubsub_v1.SubscriberClient(
            credentials=_generate_credentials())
    return SUBSCRIBER

def pull(subscription_path, max_messages=_DEFAULT_MAX_MESSAGES):
    subscriber = _get_subscriber_client()
    msgs = []
    while len(msgs) < max_messages:
        # GCP PubSub only allows us to pull 1000 messages per request,
        # so we pull until we get `max_messages` number of messages or
        # we get a response with no messages
        response = subscriber.pull(
            request={
                "subscription": subscription_path,
                "max_messages": min(1000, max_messages - len(msgs)),
            })
        if len(response.received_messages) == 0:
            break
        msgs.extend(response.received_messages)
    return msgs
```
This is some of our relevant code (the subscriber workers just repeatedly call pull until it receives zero messages from `pull()`.

One interesting theme is that in our application we also dynamically create topics and subscriptions.

Here is a picture of our metrics as well:
<img width="518" alt="image" src="https://user-images.githubusercontent.com/10524948/172635406-5839264c-5ac4-4ec7-8ee5-f61036e035f6.png">
(This is with ~500 subscribers workers)

After the subscription gets stalled like this, if I start more workers they also get very few messages before getting no messages at all (and consequently terminating).

Is this potentially a problem that affects Pub/Sub in general? Sorry if the answer to that is no and this is just a distraction, just give me the word and I can delete this comment and start a separate issue in the `google-cloud-python` repo (I hadn't seen anything related in that issue tracker). I'm in the process of boiling down our application to a minimally working example.

**Comment by coryan** [comment]
@hjtran at this point I cannot assert if this is a problem in the C++ client library or in Cloud Pub/Sub.  If I had to bet, I would say that it is more likely to be in the C++ client.

**Comment by hjtran** [comment]
Okay @coryan , is my comment useful for posterity or would you prefer that I remove it?

**Comment by coryan** [comment]
Let's keep it for now.  I am trying to repro the issue, if I am right and it is the client library, I will ping you and remove it.  Otherwise it is good info as we pass on the bug to the Pub/Sub service folks.

**Comment by coryan** [comment]
@hjtran I think this is C++ specific, I am removing your comment.  Thanks.

**Comment by coryan** [comment]
I think your code is not working due to a combination of problems:

- The client library defaults are too aggressive (they allocate too many resources) for your application
- Your application code is spending too much time performing admin operations

I have (heavily) modified your code into [a working application].  I will try to explain how this app
works below.

[a working application]: https://github.com/googleapis/google-cloud-cpp/compare/main...coryan:repro-9171

Fist some boiler plate:

```cc
#include <google/cloud/common_options.h>
#include <google/cloud/grpc_options.h>
#include <google/cloud/pubsub/publisher.h>
#include <google/cloud/pubsub/subscriber.h>
#include <google/cloud/pubsub/subscription_admin_client.h>
#include <google/cloud/pubsub/topic_admin_client.h>
#include <algorithm>
#include <atomic>
#include <iostream>

namespace g = ::google::cloud;
namespace pubsub = ::google::cloud::pubsub;
```

Then some configuration parameters.  Note the number of child topics (200) and the implied 
10 messages/second rate:

```cc
auto constexpr kTopicId = "repro-9171";
auto constexpr kDeadLetterTopicId = "repro-9171-dead-letter";
auto constexpr kSubscriptionId = "repro-9171-sub";
auto constexpr kDeadLetterSubscriptionId = "repro-9171-dead-letter-sub";
auto constexpr kChildPrefix = "child-9171-";
auto constexpr kDestinationCount = 200;
auto constexpr kMessagePeriod = std::chrono::milliseconds(100);
```

My version prints less information, but keeps counters to monitor the application:

```cc
std::atomic<std::int64_t> topic_created_count{0};
std::atomic<std::int64_t> topic_create_error_count{0};
std::atomic<std::int64_t> topic_already_exists_count{0};
std::atomic<std::int64_t> subscription_created_count{0};
std::atomic<std::int64_t> subscription_create_error_count{0};
std::atomic<std::int64_t> subscription_already_exists_count{0};
std::atomic<std::int64_t> simulation_error_count{0};
std::atomic<std::int64_t> simulation_message_count{0};
std::atomic<std::int64_t> received_count{0};
std::atomic<std::int64_t> republished_count{0};
std::atomic<std::int64_t> republished_error_count{0};
```

Then I just reused your functions to create topics and subscriptions, but these update counters
instead of logging.

```cc
void createTopic(pubsub::TopicAdminClient client, std::string const& project_id,
                 std::string const& topic_id,
                 std::string const& dead_letter_topic_id) {
  auto topic = client.CreateTopic(
      pubsub::TopicBuilder(pubsub::Topic(project_id, topic_id)));
  // Note that kAlreadyExists is a possible error when the library retries.
  if (topic.status().code() == google::cloud::StatusCode::kAlreadyExists) {
    ++topic_already_exists_count;
  } else if (!topic) {
    ++topic_create_error_count;
    throw std::runtime_error(topic.status().message());
  } else {
    ++topic_created_count;
  }

  auto dl_topic = client.CreateTopic(
      pubsub::TopicBuilder(pubsub::Topic(project_id, dead_letter_topic_id)));
  // Note that kAlreadyExists is a possible error when the library retries.
  if (dl_topic.status().code() == google::cloud::StatusCode::kAlreadyExists) {
    ++topic_already_exists_count;
  } else if (!dl_topic) {
    ++topic_create_error_count;
    throw std::runtime_error(dl_topic.status().message());
  } else {
    ++topic_created_count;
  }
}

void createOrderingSubscription(
    pubsub::SubscriptionAdminClient client, std::string const& project_id,
    std::string const& topic_id, std::string const& subscription_id,
    std::string const& dead_letter_topic_id,
    std::string const& dead_letter_subscription_id) {
  auto constexpr kDeadLetterDeliveryAttempts = 5;

  auto sub = client.CreateSubscription(
      pubsub::Topic(project_id, topic_id),
      pubsub::Subscription(project_id, subscription_id),
      pubsub::SubscriptionBuilder{}
          .enable_message_ordering(true)
          .set_ack_deadline(std::chrono::seconds(60))
          .set_dead_letter_policy(
              pubsub::SubscriptionBuilder::MakeDeadLetterPolicy(
                  pubsub::Topic(project_id, dead_letter_topic_id),
                  kDeadLetterDeliveryAttempts)));
  if (sub.status().code() == google::cloud::StatusCode::kAlreadyExists) {
    ++subscription_already_exists_count;
  } else if (!sub) {
    ++subscription_create_error_count;
    throw std::runtime_error(sub.status().message());
  } else {
    ++subscription_created_count;
  }

  auto dl_sub = client.CreateSubscription(
      pubsub::Topic(project_id, dead_letter_topic_id),
      pubsub::Subscription(project_id, dead_letter_subscription_id));
  if (dl_sub.status().code() == google::cloud::StatusCode::kAlreadyExists) {
    ++subscription_already_exists_count;
  } else if (!dl_sub) {
    ++subscription_create_error_count;
    throw std::runtime_error(dl_sub.status().message());
  } else {
    ++subscription_created_count;
  }
}
```

Here is the first interesting bit.  I am going to create a thread to send simulated data.
Note how this uses a shared `google::cloud::CompletionQueue`.  Normally the client library
creates one of these objects (and an associated thread pool) for each `pubsub::Publisher`
or `pubsub::Subcriber` created in your application.  A `CompletionQueue` is a form of
[Proactor](https://en.wikipedia.org/wiki/Proactor_pattern), an I/O demuxing mechanism.
This object also requires some file descriptors to do its work.  Normally not a problem,
but when you get into the hundreds of file descriptors you may hit the OS limits.

The other bit that is interesting is the use of an ordering key.  Ordering keys allow
you to send messages in order *as long as* they have the same ordering key. Using
different keys allow you to process messages in parallel. Given the description of
your application I assumed that ordering for messages destined to different "child
topics" was not important. If this is not a valid assumption, please let me know.

```cc
void SimulatePublisher(std::string const& project_id, g::CompletionQueue cq,
                       std::atomic<bool>& keep_running) {
  auto publisher = pubsub::Publisher(pubsub::MakePublisherConnection(
      pubsub::Topic(project_id, kTopicId),
      g::Options{}
          .set<g::GrpcCompletionQueueOption>(std::move(cq))
          .set<pubsub::MessageOrderingOption>(true)));

  auto generator = std::mt19937_64(std::random_device{}());
  auto make_random_destination = [&] {
    return std::uniform_int_distribution<int>(0,
                                              kDestinationCount - 1)(generator);
  };

  while (keep_running) {
    auto const destination = make_random_destination();
    auto const destination_topic_id =
        kChildPrefix + std::to_string(destination);
    publisher
        .Publish(pubsub::MessageBuilder{}
                     .SetOrderingKey(destination_topic_id)
                     .SetAttribute("timestamp", std::to_string(time(nullptr)))
                     .SetAttribute("topic", destination_topic_id)
                     .Build())
        .then([&](auto f) {
          ++simulation_message_count;
          auto id = f.get();
          if (!id) ++simulation_error_count;
        });
    std::this_thread::sleep_for(kMessagePeriod);
    auto const count = simulation_message_count.load();
    auto const error = simulation_error_count.load();
    if (count < 1000) continue;
    // The lack of error handling is not adequate for production, but it is
    // fine for a simulation or repro like this one.
    if (error >= count / 2) {
      std::cerr << "Too many errors in publisher thread" << std::endl;
      keep_running = false;
      break;
    }
  }
}
```

The next bit is a simple pool to keep all your `pubsub::Publisher` objects.
As I mentioned earlier in the bug, these are expensive objects to create
and destroy.  There are OS resources (file descriptors, thread pools)
created by default for each new one.  You really want to keep them
around as long as possible.  Also note that this avoids making several
admin RPCs on the handling path for each message.  Those operations are
slow, and you want to cache prior results when possible.

```cc
class PublisherPool {
 public:
  PublisherPool(std::string project_id, g::CompletionQueue cq,
                pubsub::TopicAdminClient topic_admin,
                pubsub::SubscriptionAdminClient subscription_admin)
      : project_id_(std::move(project_id)),
        cq_(std::move(cq)),
        topic_admin_(std::move(topic_admin)),
        subscription_admin_(std::move(subscription_admin)) {}
```

The important piece of this function is the caching of existing publishers:

```cc
  pubsub::Publisher MakePublisher(std::string const& topic_id) {
    std::unique_lock<std::mutex> lk(mu_);
    auto f = publishers_.find(topic_id);
    if (f != publishers_.end()) return f->second;
```

This bit is not trivial.  It releases the lock so other threads, possibly
handling other topics, can make progress. This can be important as the
admin operations (sorry to repeat myself) can be relatively slow:

```cc
    // Release the lock so other topic ids can make progress. There should not
    // be any wasted work, because the child topic id is an ordering key, a
    // single message per child topic id is processed at a time.
    lk.unlock();
    createTopic(topic_admin_, project_id_, topic_id, topic_id + "-dead-letter");
    createOrderingSubscription(subscription_admin_, project_id_, topic_id,
                               topic_id + "-sub", topic_id + "-dead-letter",
                               topic_id + "-dead-letter-sub");
    lk.lock();
    // Maybe this is overly defensive, but it is also relatively cheap. Look
    // up the publisher again
    f = publishers_.find(topic_id);
    if (f != publishers_.end()) return f->second;
```

One thing of note in this publisher, it is created with `pubsub::MessageOrderingOption(true)`.
Non-ordered publishers can batch messages and round-robin the batches across multiple I/O
sockets. That can result in re-ordered messages, which you do not want in this case.

Given the low message rate for your application, I also reduced the number of gRPC channels to `1`.
That should be enough for hundreds of messages / second, which as I understood from
your description at least an order of magnitude more than what you need.

```cc
    auto publisher = pubsub::Publisher(pubsub::MakePublisherConnection(
        pubsub::Topic(project_id_, topic_id),
        g::Options{}
            .set<pubsub::MessageOrderingOption>(true)
            .set<g::GrpcCompletionQueueOption>(cq_)
            .set<g::GrpcNumChannelsOption>(1)
            .set<pubsub::RetryPolicyOption>(
                pubsub::LimitedTimeRetryPolicy(std::chrono::minutes(5))
                    .clone())));
    publishers_.emplace(topic_id, publisher);
    return publisher;
  }

 private:
  std::string project_id_;
  g::CompletionQueue cq_;
  pubsub::TopicAdminClient topic_admin_;
  pubsub::SubscriptionAdminClient subscription_admin_;
  std::mutex mu_;
  std::unordered_map<std::string, pubsub::Publisher> publishers_;
};
```

This leaves us with `main()`.  As you do in your code, we first create
the main topic and subscription:

```cc
int main(int argc, char* argv[]) try {
  auto constexpr kUsage = R"""(Usage: <project-id> <running-time (seconds)>)""";
  if (argc != 3) throw std::runtime_error(kUsage);

  auto const project_id = std::string{argv[1]};
  auto const running_time = std::chrono::seconds(std::stoi(argv[2]));

  auto topic_admin_client =
      pubsub::TopicAdminClient(pubsub::MakeTopicAdminConnection());
  auto subscription_admin_client = pubsub::SubscriptionAdminClient(
      pubsub::MakeSubscriptionAdminConnection());

  createTopic(topic_admin_client, project_id, kTopicId, kDeadLetterTopicId);
  createOrderingSubscription(subscription_admin_client, project_id, kTopicId,
                             kSubscriptionId, kDeadLetterTopicId,
                             kDeadLetterSubscriptionId);
```

We are going to create a single completion queue and thread pool to share across all
publishers.  This is documented in [reference manual] examples (see [pub-custom], [sub-custom]). Admittedly,
the examples are not well motivated, I intend to fix that.

[reference manual]: https://googleapis.dev/cpp/google-cloud-pubsub/latest/

[pub-custom]: https://googleapis.dev/cpp/google-cloud-pubsub/latest/classgoogle_1_1cloud_1_1pubsub_1_1Publisher.html#details
[sub-custom]: https://googleapis.dev/cpp/google-cloud-pubsub/latest/classgoogle_1_1cloud_1_1pubsub_1_1Subscriber.html#details

```cc
  g::CompletionQueue cq;
  auto const pool_size = std::max(4U, std::thread::hardware_concurrency());
  std::vector<std::future<void>> tasks(pool_size);
  std::generate(tasks.begin(), tasks.end(), [cq] {
    return std::async(
        std::launch::async, [](auto cq) { cq.Run(); }, cq);
  });
```

With that custom thread pool in place we can now create the thread to simulate messages:

```cc

  std::atomic<bool> keep_running{true};
  tasks.push_back(std::async(std::launch::async, SimulatePublisher, project_id,
                             cq, std::ref(keep_running)));
```

We also create the pool of publishers, initially empty:

```cc
  PublisherPool publishers(project_id, cq, std::move(topic_admin_client),
                           std::move(subscription_admin_client));
```

And create the subscriber to receive all the events and republish them:

```cc
  auto const subscription = pubsub::Subscription(project_id, kSubscriptionId);
  auto subscriber = pubsub::Subscriber(pubsub::MakeSubscriberConnection(
      subscription, g::Options{}.set<g::GrpcCompletionQueueOption>(cq)));
  auto session = subscriber
                     .Subscribe([&](pubsub::Message m, pubsub::AckHandler h) {
                       ++received_count;
                       auto topic = m.attributes()["topic"];
                       if (topic.empty()) {
                         std::move(h).ack();
                         return;
                       }
```

The interesting bit of code here is now the handler re-publishes the event.
Note that this is more than just an optimization.  If the message is not successfully
republished the original message will be nacked, and the service will resend it.

```cc
                       publishers.MakePublisher(topic)
                           .Publish(std::move(m))
                           .then([h = std::move(h)](auto f) mutable {
                             ++republished_count;
                             auto id = f.get();
                             if (!id) {
                               ++republished_error_count;
                               std::move(h).nack();
                               std::cerr << "Problem re-publishing message "
                                         << id.status() << std::endl;
                               return;
                             }
                             std::move(h).ack();
                           });
                     })
                     .then([&](auto f) {
                       keep_running = false;
                       return f.get();
                     });
```

Next, we run the program for the prescribed time, and print out progress every
few seconds:

```cc
  auto now = [] { return std::chrono::steady_clock::now(); };
  auto deadline = now() + running_time;
  while (keep_running && now() < deadline) {
    auto const state = session.wait_for(std::chrono::seconds(5));
    if (state != std::future_status::timeout) keep_running = false;
    std::cerr << "\nReceived count: " << received_count
              << "\nSimulation count: " << simulation_message_count
              << "\nSimulation errors: " << simulation_error_count
              << "\nCreated topics: " << topic_created_count
              << "\nAlready existing topics: " << topic_already_exists_count
              << "\nCreate topic errors: " << topic_create_error_count
              << "\nCreated subscriptions: " << subscription_created_count
              << "\nAlready existing subscriptions: "
              << subscription_already_exists_count
              << "\nCreate subscription errors: "
              << subscription_create_error_count << std::endl;
  }
```

Finally, we shut down the completion queue and wait for all running tasks:

```cc
  keep_running = false;
  session.cancel();
  std::cerr << "Session finishes with status " << session.get() << "\n";

  cq.Shutdown();
  for (auto& t : tasks) t.get();
  return 0;
} catch (std::exception const& ex) {
  std::cerr << "Standard exception thrown: " << ex.what() << "\n";
  return 1;
}
```

**Comment by coryan** [comment]
> I'm finding the documentation to be not the best (I didn't see anything about RetryPolicyOption on the web page that discusses error handling) and the only sample code I could find is a single file that executes all the functions.

There is an example on how to set the `RetryPolicy` and `BackoffPolicy` here:

https://googleapis.dev/cpp/google-cloud-pubsub/latest/classgoogle_1_1cloud_1_1pubsub_1_1Subscriber.html#details

Your comment does point out that this is not well motivated, it does not say why you may want to set it, or what it controls. And you are correct that this example is not surfaced in `cloud.google.com/pubsub/docs`.  It is very specific to C++, but we should figure out a way to put it out there.

There is also some missing documentation on the internals of the C++ Pub/Sub client library, something that gives you a sense of what resources are created by a `pubsub::Publisher` and/or `pubsub::Subscriber`.

**Comment by BrianPez** [comment]
I've made changes based on your example and have deployed to my test environment.

Couple of things as I worked on integrating your changes.
1) My code needs to be self correcting.  In other words, if the subscription fails, I need the process to auto-recreate the subscription and keep running, not exit or just stop subscribing.  This is another reason for the original double while() loop.
2) It is possible for the other thread that populates my outgoing message queue to send messages before the PubSub thread is ready. I need the outgoing message queue to make sure messages aren't lost before the PubSub thread is ready to send them.

**Comment by BrianPez** [comment]
After 3 1/2 days, one of my test sites stopped receiving messages.  Process started 6/10 @ 14:40:02.  Last message received 6/14 @ 7:50:55.  2 subscription topics and subscriptions (1 regular & 1 dead letter) for created 6/10.  4 publish topics and subscriptions (2 regular & 2 dead letter) were created on 6/10.

The failed site is much slower over the weekends.  Subscription receive message count for the full days:
Jun11: 6000
Jun12: 5993
Jun13: 8176
Publish message counts:
Jun11: 2
Jun12: 2
Jun13: 82

This morning, subscription receive got 2372 messages before the problem and published 0 messages.  It was in the problem state for about 10 minutes before I restarted.  I received about 100 subscribe messages in the first 10 minutes of processing (278 total as of my typing) and published 24 messages.

No error message from the .then() handler on the subscriber.  The inner while() loop was never exited.  Here is the current code:
`
    void createTopicSubscriptionId(size_t const& enterpriseId, size_t const& facilityId, std::string const& gatewayId, std::string& topicId, std::string& subscriptionId, std::string& dlTopicId, std::string& dlSubscriptionId)
    {
        if (enterpriseId > 0 && facilityId > 0)
        {
            char hostname[1024];

            topicId = subscriptionId =  TOPIC_ID_PREFIX_FIELDNETD;
            topicId += "t-";
            subscriptionId += "s-";
            topicId += std::to_string(enterpriseId) + "-" + std::to_string(facilityId);
            subscriptionId += std::to_string(enterpriseId) + "-" + std::to_string(facilityId);
            gethostname(hostname, 1024);
            if (strcmp(hostname, "brian-lt") == 0)
            {
                topicId += "-brian-lt";
                subscriptionId += "-brian-lt";
             }
        }
        else if (! gatewayId.empty())
        {
            topicId = subscriptionId = TOPIC_ID_PREFIX_GATEWAY;
            topicId += "t-";
            subscriptionId += "s-";
            topicId += gatewayId;
            subscriptionId += gatewayId;
        }
        else
        {
            LOGc << "Unsupported queue type, exiting";
            exit(EXIT_UNKNOWN_QUEUE_TYPE);
        }
        dlTopicId = topicId + "-dl";
        dlSubscriptionId = subscriptionId  + "-dl";
        LOGn << "topic/subscription is " << topicId << "/" << subscriptionId;
    }
    void createTopic(pubsub::TopicAdminClient client,
                     std::string const& project_id,
                     std::string const& topic_id,
                     std::string const& dead_letter_topic_id)
    {
        auto topic = client.CreateTopic(pubsub::TopicBuilder(pubsub::Topic(project_id, topic_id)));
        // Note that kAlreadyExists is a possible error when the library retries.
        if (topic.status().code() == google::cloud::StatusCode::kAlreadyExists)
        {
            ;
        }
        else if (!topic)
        {
          throw std::runtime_error(topic.status().message());
        }
        else
        {
            LOGn << "The topic " << topic_id << " was successfully created: " << topic->DebugString();
        }

        auto dl_topic = client.CreateTopic(
            pubsub::TopicBuilder(pubsub::Topic(project_id, dead_letter_topic_id)));
        // Note that kAlreadyExists is a possible error when the library retries.
        if (dl_topic.status().code() == google::cloud::StatusCode::kAlreadyExists)
        {
           ;
        }
        else if (!dl_topic)
        {
          throw std::runtime_error(dl_topic.status().message());
        }
        else
        {
            LOGn << "The topic " << dead_letter_topic_id << " was successfully created: " << topic->DebugString();
        }
    }
    void createOrderingSubscription(pubsub::SubscriptionAdminClient client,
                                    std::string const& project_id,
                                    std::string const& topic_id,
                                    std::string const& subscription_id,
                                    std::string const& dead_letter_topic_id,
                                    std::string const& dead_letter_subscription_id)
    {
        auto constexpr kDeadLetterDeliveryAttempts = 5;

        auto sub = client.CreateSubscription(pubsub::Topic(project_id, topic_id),
                                             pubsub::Subscription(project_id, subscription_id),
                                             pubsub::SubscriptionBuilder{}
                                                 .enable_message_ordering(true)
                                                 .set_ack_deadline(std::chrono::seconds(60))
                                                 .set_dead_letter_policy(
                                                     pubsub::SubscriptionBuilder::MakeDeadLetterPolicy(
                                                         pubsub::Topic(project_id, dead_letter_topic_id),
                                                         kDeadLetterDeliveryAttempts)
                                                 )
        );
        if (sub.status().code() == google::cloud::StatusCode::kAlreadyExists)
        {
            ;
        }
        else if (!sub)
        {
            throw std::runtime_error(sub.status().message());
        }
        else
        {
            LOGd << "The subscription " << subscription_id << " was successfully created: " << sub->DebugString();
        }

        auto dl_sub = client.CreateSubscription(pubsub::Topic(project_id, dead_letter_topic_id),
                                                pubsub::Subscription(project_id, dead_letter_subscription_id));

        if (dl_sub.status().code() == google::cloud::StatusCode::kAlreadyExists)
        {
            ;
        }
        else if (!dl_sub)
        {
            throw std::runtime_error(dl_sub.status().message());
        }
        else
        {
            LOGd << "The deadletter subscription " << dead_letter_subscription_id << " was successfully created: " << dl_sub->DebugString();
        }
    }
      class PublisherPool
      {
      private:
          std::string project_id_;
          pubsub::TopicAdminClient topic_admin_;
          pubsub::SubscriptionAdminClient subscription_admin_;
          std::mutex mu_;
          std::unordered_map<std::string, pubsub::Publisher> publishers_;

      public:
          PublisherPool(std::string project_id,
                        pubsub::TopicAdminClient topic_admin,
                        pubsub::SubscriptionAdminClient subscription_admin)
            : project_id_(std::move(project_id)),
              topic_admin_(std::move(topic_admin)),
              subscription_admin_(std::move(subscription_admin)) {}

          pubsub::Publisher MakePublisher(std::string const& topic_id,
                                          std::string const& subscription_id,
                                          std::string const& dead_letter_topic_id,
                                          std::string const& dead_letter_subscription_id)
          {
              std::unique_lock<std::mutex> lk(mu_);
              auto f = publishers_.find(topic_id);
              if (f != publishers_.end()) return f->second;
              // Release the lock so other topic ids can make progress. There should not
              // be any wasted work, because the child topic id is an ordering key, a
              // single message per child topic id is processed at a time.
              lk.unlock();
              GCPPubSubController::createTopic(topic_admin_, project_id_, topic_id, dead_letter_topic_id);
              GCPPubSubController::createOrderingSubscription(subscription_admin_,
                                                              project_id_,
                                                              topic_id,
                                                              subscription_id,
                                                              dead_letter_topic_id,
                                                              dead_letter_subscription_id);
              lk.lock();
              // Maybe this is overly defensive, but it is also relatively cheap. Look
              // up the publisher again
              f = publishers_.find(topic_id);
              if (f != publishers_.end()) return f->second;
              auto publisher = pubsub::Publisher(pubsub::MakePublisherConnection(
                  pubsub::Topic(project_id_, topic_id),
                  ::google::cloud::Options{}
                      .set<pubsub::MessageOrderingOption>(true)
                      .set<::google::cloud::GrpcNumChannelsOption>(1)
                      .set<pubsub::RetryPolicyOption>(pubsub::LimitedTimeRetryPolicy(std::chrono::minutes(5)).clone())
              ));
              publishers_.emplace(topic_id, publisher);
              return publisher;
            }
      };
void GCPPubSubController::run(const size_t& e, const size_t& f)
{
    while(running)
    {
        try
        {
            bool subscribeError = false;
            auto topic_admin_client = pubsub::TopicAdminClient(pubsub::MakeTopicAdminConnection());
            auto subscription_admin_client = pubsub::SubscriptionAdminClient(pubsub::MakeSubscriptionAdminConnection());

            // Create the incoming ID strings
            createTopicSubscriptionId(e,
                                      f,
                                      "",
                                      fTopicId,
                                      fSubscriptionId,
                                      fDLTopicId,
                                      fDLSubscriptionId);
            // Make sure the incoming topic exists
            createTopic(topic_admin_client, projectId, fTopicId, fDLTopicId);
            // Makre sure the outging topic exists
            createOrderingSubscription(subscription_admin_client,
                                       projectId,
                                       fTopicId,
                                       fSubscriptionId,
                                       fDLTopicId,
                                       fDLSubscriptionId);
            // Create the pool for outgoing messages
            PublisherPool publishers(projectId, std::move(topic_admin_client), std::move(subscription_admin_client));

            // Subscribe for incoming messages
            auto const subscription = pubsub::Subscription(projectId, fSubscriptionId);
            auto subscriber = pubsub::Subscriber(pubsub::MakeSubscriberConnection(subscription));
            auto session = subscriber.Subscribe([&](pubsub::Message m, pubsub::AckHandler h) {
                std::string msg = m.data();

                LOGi << "Received message, id " << m.message_id() << ", attempt " << h.delivery_attempt() <<", ordering key '" << m.ordering_key() << "', topic " << ProcessMsg::commTopic(topic) << " (" << topic << ")";
                std::move(h).ack();
            })
            .then([&](::google::cloud::future<google::cloud::Status> f) {
                LOGe << "Subscription problem : " << f.get();
                subscribeError = false;
                // Causes an abort after a SIGTERM and shutdown in progress
                //return f.get();
                return "";
            });

            while (! subscribeError)
            {
                if (! running)
                {
                    LOGi << "Cancelling session";
                    session.cancel();
                    break;
                }

                if (! outgoingQueue.empty())
                {
                    std::string gTopicId = ""; 
                    std::string gSubscriptionId = ""; 
                    std::string gDLTopicId = ""; 
                    std::string gDLSubscriptionId = ""; 
                    const OutgoingMsg omsg = outgoingQueue.front();

                    // Make sure the Pub/Sub topic has been created
                    createTopicSubscriptionId(0, 0, omsg.identifier, gTopicId, gSubscriptionId, gDLTopicId, gDLSubscriptionId);
                    auto id = publishers.MakePublisher(gTopicId,
                                                       gSubscriptionId,
                                                       gDLTopicId,
                                                       gDLSubscriptionId)
                                  .Publish(std::move(pubsub::MessageBuilder{}
                                                         .SetData(omsg.msg)
                                                         .SetData(omsg.msg)
                                                         .SetAttribute("msgCounter", std::to_string(++msgCounter))
                                                         .SetAttribute("timestamp", std::to_string(time(NULL)))
                                                         .Build()))
                                  .get();
                    if (! id)
                    {
                        LOGe << "Failed to publish message " << id.status().message();
                    }
                    else
                    {
                        LOGd << "Published message, id : " << *id << " (" << omsg.msg << ") to " << projectId << "/" << gTopicId;
                        outgoingQueue.pop_front();
                    }
                }

                auto const state = session.wait_for(std::chrono::milliseconds(500));
                if (state != std::future_status::timeout)
                {
                    LOGe << "Subscribe error on wait_for : " << session.get();
                    subscribeError = false;
                }
            } // while (! subscribeError)
            if (! running)
            {
                break;
            }

            LOGn << "Subscribe stopped with status " << session.get();
            session.cancel();
            LOGd << "Waiting 15 seconds";
            std::this_thread::sleep_for(std::chrono::seconds(15));
        }
        catch (std::runtime_error &e)
        {
            LOGe << "RuntimeError exception : " << e.what();
        }
        catch(std::exception& e)
        {
            LOGe << "Exception : " << e.what();
        }
    } // while (running)
}
`

**Comment by coryan** [comment]
I think we are going to have to look at the logs on the server side.  Do you mind opening a support ticket:

https://cloud.google.com/support/

I have no access to the service logs, without a support ticket there is very little I can do here.

**Comment by BrianPez** [comment]
Unless I am misreading the web page you sent, it appears the only way to submit a ticket is to pay for standard support at $29 per month?

**Comment by coryan** [comment]
Candidly, I have never used support myself.  I have helped the support folks, but I am not familiar with the billing policies.  I assume your reading of the site is correct.


I took a second look at your code.  It is missing some changes that I suggested:


> Here is the first interesting bit. I am going to create a thread to send simulated data.
> Note how this uses a shared `google::cloud::CompletionQueue`. Normally the client library
> creates one of these objects (and an associated thread pool) for each `pubsub::Publisher`
> or `pubsub::Subcriber` created in your application. A `CompletionQueue` is a form of
> [Proactor](https://en.wikipedia.org/wiki/Proactor_pattern), an I/O demuxing mechanism.
> This object also requires some file descriptors to do its work. Normally not a problem,
> but when you get into the **hundreds of file descriptors you may hit the OS limits**.

Is there a reason you decided not to share a completion queue across all the publishers and subscribers?  I think that was an important piece of the puzzle.

**Comment by BrianPez** [comment]
My incoming and outgoing messages have no relation to each other.  I had thought the completion queue was part of your test program used to reflect incoming messages out to the publishers.  To be honest, I find the Google PubSub C++ implementation extremely complicated, especially when compared to AWS SQS or ZeroMQ.  As I said previously, the documentation seems very sparse.  A Google search on CompletionQueue only turns up the class reference.

**Comment by coryan** [comment]
Were you able to test with a shared completion queue?  Did that help?

**Comment by BrianPez** [comment]
I have not.  I’m not sure when I will have a change to try implementing it.

**Comment by coryan** [comment]
> I have not. I’m not sure when I will have a change to try implementing it. 

Fair enough.  I have created some bugs to improve the documentation gaps you pointed out.  It is also evident that we need a simpler API (probably in addition to the existing one).  Our current APIs are designed to help with high throughput applications, but obviously they are a hindrance to more normal appls.  I will close this issue for now.  If you run into problems again, please feel free to reopen it, or create a new one, at your convenience.
